GoodCigarFilter + tests

akiezun opened the pull request:
ported BadCigarFilter (in Hellbender filters have positive names, in concordance with Java filter semantics) + tests (added 2 tests to cover 2 more branches, 1 branch seems unreachable).

@vruano  please review
addresses https://github.com/broadinstitute/hellbender/issues/373

akiezun commented:
@vruano is on vacation. @lbergelson please review

vruano commented:
Done with my review. I'm to assigning it back to @akiezun in case @lbergelson wants to add.

akiezun commented:
I think we should write most readable code for now and optimize with a profiler later. The regexp machinery in Java libraries definitely does a better job optimizing the state machine than a person can.

vruano commented:
Finished with my review back to @akiezun

vruano commented:
Please check on my general comments posted at the end of the second commit, I thought that they would be posted here as well.

akiezun commented:
@vruano you're raising 3 issues:
1) is the algorithm correct?
2) is the algorithm's results the same as in GATK3?
3) is the algorithm esthetically pleasing/readable.

For 1, I'm going to create a new ticket - it's not under discussion here.
For 2, I'll enter the cases you're suggesting as unit tests and see what happens.
For 3, I can work on making it read better. 

My first goal in this tickets is to port the existing filter. If we can improve the readability at the same time, that's great. Fixing any bugs or deficiencies of the algorithm is strictly future work (if it has been broken all this time in GATK3, it's ok for it to be broken for a few more weeks in GATK4).

akiezun commented:
@eitanbanks and @vdauwera I heard from @droazen that you guys are working on the BadCigarFilter in GATK3. Can you summarize the work? we need to synchronize the changes/improvements across the 2 codebases.

eitanbanks commented:
BadCigarFilter was made a default Walker filter and functionality was added to it so that it now also filters out reads whose cigar length does not equal the read length.

akiezun commented:
is the code in?

On Tue, Apr 21, 2015 at 4:51 PM, Eric Banks <notifications@github.com>
wrote:

> BadCigarFilter was made a default Walker filter and functionality was
> added to it so that it now also filters out reads whose cigar length does
> not equal the read length.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/hellbender/pull/380#issuecomment-94937929>
> .
>

vdauwera commented:
See https://github.com/broadinstitute/gsa-unstable/pull/930

The title is misleading -- see the last few comments in the conversation for a summary of the work done.

akiezun commented:
why not rename the issue then?

akiezun commented:
@eitanbanks thanks. the code and tests are now in hellbender too.

akiezun commented:
@vruano can you give an example for "Also it does accept non empty CIGAR with out any operator that consume read bases such as N or P. That would not happen with the GATK3 code." ?

akiezun commented:
@vruano back to you

I fixed the 1H1H1M problem you brought up "GATK3's filter would not complain for consecutive H operations at the beginning whereas the new code does" and I entered tests for it (one for head, one for tail).

I entered https://github.com/broadinstitute/hellbender/issues/428 for clarifying the spec.

vruano commented:
What is that for example "10P" or "101N" would fail in the GATK3 code but the would not in the current pull-request. The reason is that the P and N operators return false for the method "consumesReadBases()" and the GATK3 code pays attention to that.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/filters/GoodCigarFilter.java:
```diff
@@ -0,0 +1,83 @@
+ */
+public final class GoodCigarFilter implements ReadFilter {
+
+    private static final String DIGITS = "\\d*"; //Any operator can be preceeded by a series of digits.
```
This is incorrect -- each operator must be preceded by a series of digits. The regexp for an entire CIGAR string (from the SAM spec) is: `\*|([0-9]+[MIDNSHPX=])+`

akiezun replied:
True. A cigar string without a number is malformed so you can't make a samrecord. In the regexp setting you can at least test the string itself and validate that if there ever were a read with this cigar we would reject it. Anyway, fixed.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/filters/GoodCigarFilter.java:
```diff
@@ -0,0 +1,83 @@
+               (ANYTHING + any(I+I, D+D, I+D, D+I) + ANYTHING),                //consecutive indels
+               (maybeSome(CLIP)+ D + ANYTHING),                                //starts with deletion
+               (ANYTHING + D + maybeSome(CLIP)),                               //ends with deletion
+               (some(CLIP))));                                                 //only clips
```
This approach makes me incredibly nervous -- complex regular expressions are notoriously prone to edge conditions and hard-to-see bugs, especially when (as here) there are multiple indefinite quantifiers. My experience has been that regular expressions with more than 2-3 `*` and/or `+` quantifiers tend to be hopelessly brittle and prone to weird results. Eg., are you ABSOLUTELY sure that you want to be always using greedy quantifiers (`*` and `+`) instead of reluctant quanitifiers (`*?` and `+?`) in your `maybeSome()` and `some()` methods? Have you thought about how these interact in every scenario? I can almost guarantee that there is at least one case where a `*` or `+` will eat too much of the cigar and lead to some erroneous result...

I feel that a straightforward parsing of the Cigar elements would be far less error-prone.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/filters/GoodCigarFilter.java:
```diff
@@ -0,0 +1,83 @@
+
+    //We pre-create one regular expression pattern and reuse it.
+    //Each clause in the pattern corresponds to one condition we're testing for.
+    private final static Pattern BAR_CIGAR = Pattern.compile(
```
Should be named `BAD_CIGAR`, not `BAR_CIGAR`

vruano commented:
Also it seems that several soft clips are allowed in GATK3 code as long as is before or after the non-clipping core section of the CIGAR.

So the following would pass the filter:

1S20S10M 
1H1S10M10S30H
1H1S10M10S1S30H

But these wouldn't:

1S1H10M
1H1M1S1M
1M1H1S

vruano commented:
Is much the same as with H.

vruano commented:
Done with the review.

akiezun commented:
@vruano to clarify. both 10P and 101N pass on GATK3. I just checked.

akiezun commented:
by "pass" i mean - they are not flagged as Bad

akiezun commented:
alright @vruano @droazen  
as discussed today I will recode this as a bunch of loops - the GATK3 code is atrocious but my regexp did not me win the fan following i was going for. And ultimately it's about the readability at this point (which is a subjective manner, you guys said loops would make it more readable and people's opinion defines 'readability')

vruano commented:
"@vruano to clarify. both 10P and 101N pass on GATK3. I just checked."

I see... it seems that this might be cause by yet another bug in GATK3 code ... but I guess that is to be left for the other issue.

From this code within the second loop:

```java
if (!hasMeaningfulElements && op.consumesReadBases()) {
                    hasMeaningfulElements = true;
}
```
one would think that to have a meaningful-element is to have one that consumes read bases.

However it turns out that this if condition is never true as:

```java
boolean hasMeaningfulElements = (firstOp != CigarOperator.H && firstOp != CigarOperator.S);
```
is guaranteed to initialize it to true. 

So that is some code you can get rid off for now. But I guess that it need to be fixed eventually.

akiezun commented:
Does anyone know why there are 3 implementation of validity for Cigars in GATK3?
1) in the BadCigarFilter
2) in CigarUtils
3) in Cigar itself ?

akiezun commented:
the third one is in htsjdk obviously. anyway, i'm merging the first two

akiezun commented:
I clarified the semantics of Good CIGAR. It does not make sense to have 'good cigars' that are invalid according to htsjdk so I clarified the semantics to state that a "good" cigar is a valid cigar that passes additional criteria:
```
  - does not have consecutive I/D elements
  - does not start or end with deletions (with or without preceding clips).
```
(now, we can argue about those criteria but those are the criteria in GATK3 and the current code and doc makes this explicit.)

I rewrote the whole code again to reflect this change. @vruano back to you.

vruano commented:
The code is far easier to understand now. Mostly minor changes and a suggestion to improve the readability of one of the supporting private methods. 

CigarUtils.countRefBasesBasedOnCigar have some issues but these are not really part of your code so it might be fair to address them separately in another pull-request. 

Back to @akiezun.

akiezun commented:
This filter did not deserve a special file after I moved the isGood method to CigarUtils. So I moved this filter to ReadFilterLibrary and moved tests accordingly.

back to @vruano

vruano commented:
@akiezun please rebase, squash and merge.
